---
title: "IDGenomic-vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{IDGenomic-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This is a package to process data loaded directly from [Gene Expression Omnibus](https://www.ncbi.nlm.nih.gov/geo/), then ... 

```{r setup}
library(IDGenomic)
```

```{r, message = FALSE}
library(SummarizedExperiment)
```

# Extracting data from GEO

Currently, this package supports parsing GSE matrix from GEO. 

GSE48762 is a dataset related to vaccination and is part of the GSE30101 series. The study involved participants who received either a flu vaccine, pneumococcal vaccine, or a saline solution. This dataset provides valuable information for researchers interested in the immune response to different vaccines and can contribute to the development of more effective vaccination strategies. Metadata and demographic information are deposited in GEO.

#### Input for GEOExtract()

- `GEOAccession` must be provided as a character.

- `GSEMatrix` is a boolean indicating whether to extract the GSE matrix from GEO, with a default value of TRUE.

- `getGPL` is an option that comes from the package [GEOquery::getGEO](https://www.rdocumentation.org/packages/GEOquery/versions/2.38.4/topics/getGEO), with a default value of TRUE.

- `plots` is an option that will output a boxplot of 10 randomly selected samples, with a default value of TRUE.

- `NSub` is an option that specifies the number of samples to be randomly selected for the boxplot, with a default value of 10.

#### Output from GEOExtract()

The function will print information related to the GEO dataset, specifically information included in featureData and phenoData.

If `plots = TRUE`, a boxplot will be generated for `NSub` randomly selected samples.

#### Return from GEOExtract()

One [SummerizedExperiment](https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html) class object contains both expression and metadata related to each sample and feature.

```{r, fig.width=7, fig.height=5}
GSE48762 <- GEOExtract(GEOAccession = "GSE48762",
                       GSEMatrix =TRUE, 
                      getGPL = TRUE,  
                      plots = TRUE, 
                      NSub = 10)

```

# Examine GEO data

Upon examining the result from `GeoExtract()`, gene symbol is provided when we apply `getGPL = TRUE` option. We will examine `c("Symbol", "Entrez_Gene_ID")` from featureData.

Upon examining the results from `GeoExtract()`, gene symbols are provided when we apply the `getGPL = TRUE` option. We can examine `c("Symbol", "Entrez_Gene_ID")` in featureData to view the gene symbols and Entrez gene IDs associated with the features in the dataset.


```{r}
head(rowData(GSE48762)$Symbol)
head(rowData(GSE48762)$Entrez_Gene_ID)
```

Further, we are interested in specific characteristics related to each sample. We will investigate the following metadata fields: `c("blood source:ch1", "day:ch1", "ethnicity:ch1", "gender:ch1", "vaccine:ch1")`.

```{r}
table(GSE48762$`blood source:ch1`)

table(GSE48762$`day:ch1`)

table(GSE48762$`ethnicity:ch1`)

table(GSE48762$`gender:ch1`)

table(GSE48762$`vaccine:ch1`)
```

# Quantile Normalization

Based on the boxplot and information provided in the dataset, we will perform quantile normalization. Function in `IDGenomic` for quantile normalization is called from package `limma`.

#### Input for QuantileNorm()

- `GEOdata` is a matrix or SummerizedExperiment object. If a SummarizedExperiment object is provided, it already contains the expression data and its corresponding metadata. If a matrix is provided, rows are features and columns are samples (this structure is maintained throughout the entire package). 

- `plots` option generates a boxplot of 10 (default) randomly selected samples, this option is set to TRUE by default. 

- `NSub` option allows you to specify the number of samples to be randomly selected for the boxplot when the `plots` option is set to TRUE. `NSub = 10` is set by default.

### Output from QuantileNorm()

If `plots = TRUE`, a boxplot will be generated for `NSub` randomly selected samples.

### Return from QuantileNorm()

A SummerizedExperiment object contains both quantile-normalized expression and metadata related to each sample, or a matrix of quantile-normalized expression only. For a SummerizedExperiment object, quantile-normalized data is stored in `assays(GEOdata)$quantileNorm`, while the original expression data is preserved.

```{r, fig.width=7, fig.height=5}
class(GSE48762)

GSE48762 <- QuantileNorm(GEOdata = GSE48762, plots = TRUE, NSub = 10)
```

# Subset features

This function allows for the subsetting of a pre-defined list of probes/genes, with the option to select unique probes/genes. It is not necessary for the data to be quantile normalized.

#### Input for FeatureFilter()

- `GEOdata` is either a matrix or SummerizedExperiment object. If a SummerizedExperiment object is provided, GEOdata contains both expression and metadata related to each sample and feature. If a matrix is provided, GEOdata contains expression data, and the `rownames(GEOdata)` must correspond to the `filterValue`.

- `filterVar` should be a character indicating the variable that is being subsetted related to the feature of expression data for SummerizedExperiment object. The default value for filterVar is NA. Note that if a matrix is provided as GEOdata, there is no need to provide `filterVar`.

- `filterValue` is a vector of gene/probe identities that one wishes to keep in the dataset after subsetting. If a SummerizedExperiment object is provided, `filterValue` is a subset of values in `filterVar`. If a matrix is provided as GEOdata, `filterValue` is a subset of `rownames(GEOdata)`.

- `unique` parameter is a boolean that determines whether to keep only unique genes/probes after filtering. Its default value is FALSE. If `unique=TRUE`, only unique genes/probes are retained by selecting the gene/probe with the highest median expression level.

#### Return from FeatureFilter()

A SummerizedExperiment object or matrix.

```{r}
dim(GSE48762)

gene_overlap <- openxlsx::read.xlsx("/Users/yu.zhang/Desktop/Modular/Data/total_overlap_8.24.xlsx", sheet = 1)$with_U133A
gene_overlap <- gene_overlap[!is.na(gene_overlap)]
length(gene_overlap)

GSE48762.sub <- FeatureFilter(GEOdata = GSE48762, 
                              filterVar = "Symbol",
                              filterValue = gene_overlap,
                              unique = FALSE)
dim(GSE48762.sub)

GSE48762.unique <- FeatureFilter(GEOdata = GSE48762, 
                              filterVar = "Symbol",
                              filterValue = gene_overlap,
                              unique = TRUE)
dim(GSE48762.unique)

# If all unique genes, should return FALSE.
all(duplicated(rowData(GSE48762.unique)$Symbol))
```


## When gene symbols cannot be retrived directly from GEO

Annotation of probes to gene symbol will be done through BioMart packge. Function/code/demonstration will be added.




