---
title: "IDGenomic-vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{IDGenomic-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This is a package to process data loaded directly from [Gene Expression Omnibus](https://www.ncbi.nlm.nih.gov/geo/), then ... 

```{r setup}
library(IDGenomic)
```

# Extracting data from GEO

Currently, this package supports parsing GSE matrix from GEO. 

GSE48762 is a vaccine data, part of the GSE30101 series. Participants received either a Flu vaccine, pnuemococcoal vaccine or saline. 

#### Input for GeoExtract()

- `GEOAccession` must be provided as a character.

- `getGPL` options comes from package [GEOquery::getGEO](https://www.rdocumentation.org/packages/GEOquery/versions/2.38.4/topics/getGEO). (default is TRUE)

- `plots` option will output a boxplot of 10 (default) randomly selected samples. (default is TRUE)

- `NSub` option specify number of samples to be randomly selected for boxplot. (default is 10)

#### Output from GeoExtract()

Information related to the GEO dataset will be printed. Especially inforamtion included in featureData and phenoData are listed.

If `plots = TRUE`, then a boxplot will be drawn for `NSub` number of randomly selected samples.

#### Return from GeoExtract()

One [SummerizedExperiment](https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html) class object contains both expression and metadata related to each sample and feature.

```{r, fig.width=7, fig.height=5}
GSE48762 <- GeoExtract(GEOAccession = "GSE48762",
                      getGPL = TRUE,  
                      plots = TRUE, 
                      NSub = 10)

```

# Examine GEO data

Upon examining the result from `GeoExtract()`, gene symbol is provided when we apply `getGPL = TRUE` option. We will examine `c("Symbol", "Entrez_Gene_ID")` from featureData.

```{r}
head(rowData(GSE48762)$Symbol)
head(rowData(GSE48762)$Entrez_Gene_ID)
```

Further, we are interested in certain characteristics regarding each sample. We will look into `c("blood source:ch1", "day:ch1", "ethnicity:ch1", "gender:ch1", "vaccine:ch1")`.

```{r}
table(GSE48762$`blood source:ch1`)

table(GSE48762$`day:ch1`)

table(GSE48762$`ethnicity:ch1`)

table(GSE48762$`gender:ch1`)

table(GSE48762$`vaccine:ch1`)
```

# Quantile Normalization

From the boxplot and description given in this data, we will perform quantile normalization. Function for quantile normalization is called from `limma()`.

#### Input for QuantileNorm()

- `GEOdata` either a matrix or SummerizedExperiment object contains expression. If a matrix is provided, rows are features and columns are samples (this structure remains for the entire pacakge).

- `plots` option will output a boxplot of 10 (default) randomly selected samples. (default is TRUE)

- `NSub` option specify number of samples to be randomly selected for boxplot. (default is 10)

### Output from QuantileNorm()

If `plots = TRUE`, then a boxplot will be drawn for `NSub` number of randomly selected samples after quantile normalization.

### Return from QuantileNorm()

Either SummerizedExperiment object contains both quantile normalized expression and metadata related to each sample, or a matrix of expression only.

Quantile normalized data is stored in `assays(GEOdata)$quantileNorm` for SummerizedExperiment object, and original expression data is kept.

```{r, fig.width=7, fig.height=5}
class(GSE48762)

GSE48762 <- QuantileNorm(GEOdata = GSE48762, plots = TRUE, NSub = 10)
```

# Subset features

Given a pre-defined list of probes/genes, this function subset to this list of probes/genes and give the option to pick unique probes/genes. Data does not need to be quantile normalized.

#### Input for FeatureFilter()

- `GEOdata` either a matrix or SummerizedExperiment object contains expression. If a matrix is provided, `rownames(GEOdata)` must be the `filterValue`.

- `filterVar` a character for the variable that is being subsetted related to the feature of expression data (default = NA). If a matrix is provided, user does not need to provide `filterVar`.

- `filterValue`  a vector of genes/probes identities one wish to keep

- `unique` a boolean, whether to keep only unique genes/probes after filter (default = FALSE). If `unique=TRUE`, probes/genes with the highest median expression is kept.

#### Return from FeatureFilter()

A matrix or SummerizedExperiment object.

```{r}
dim(GSE48762)

gene_overlap <- openxlsx::read.xlsx("/Users/yu.zhang/Desktop/Modular/Data/total_overlap_8.24.xlsx", sheet = 1)$with_U133A
gene_overlap <- gene_overlap[!is.na(gene_overlap)]
length(gene_overlap)

GSE48762.sub <- FeatureFilter(GEOdata = GSE48762, 
                              filterVar = "Symbol",
                              filterValue = gene_overlap,
                              unique = FALSE)
dim(GSE48762.sub)

GSE48762.unique <- FeatureFilter(GEOdata = GSE48762, 
                              filterVar = "Symbol",
                              filterValue = gene_overlap,
                              unique = TRUE)
dim(GSE48762.unique)

# If all unique genes, should return FALSE.
all(duplicated(rowData(GSE48762.unique)$Symbol))
```


## When gene symbols cannot be retrived directly from GEO

Annotation of probes to gene symbol will be done through BioMart packge. Function/code/demonstration will be added.




